master/internal/rm/kubernetesrm/jobs.go

```diff
@@ -1811,6 +1824,54 @@ func allTaintsTolerated(taints []k8sV1.Taint, tolerations []k8sV1.Toleration) bo
 	return true
 }
 
+// Check that pods belong to this resource pool (from poolTCD) can be scheduled on a given node.
+func (j *jobsService) podsCanBeScheduledOnNode(affinity *k8sV1.NodeSelector, node *k8sV1.Node) bool {
+	// In case of no defined affinities/node selectors, the pod can default to schedule on the given node.
+	if affinity == nil || affinity.NodeSelectorTerms == nil {
+		return true
+	}
+	// If there are defined node selectors but the node doesn't have any labels, return false.
+	if node.Labels == nil {
+		return false
+	}
+
+	ns, err := nodeaffinity.NewNodeSelector(affinity)
+	if err != nil {
+		j.syslog.WithError(err)
+		return false
+	}
+	return ns.Match(node)
+}
+
+// Gets the node affinity/selector from the resource pool.
+func extractNodeSelectors(pod *k8sV1.Pod) *k8sV1.NodeSelector {
+	if pod == nil {
+		return nil
+	}
+
+	affinity := &k8sV1.NodeSelector{}
+
+	// First check for node affinities.
+	if nodeAffinity := pod.Spec.Affinity; nodeAffinity != nil {
+		affinity = nodeAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+	}
+
+	// Then add any node labels.
+	if nodeSelector := pod.Spec.NodeSelector; nodeSelector != nil {
```

Comment:
have you tested a case where both node affinity and a node selector is specified?

I think adding them to this list will make the requirement ```ORed``` when I think we want the requirement to be ```ANDed```?
```
// A node selector represents the union of the results of one or more label queries
// over a set of nodes; that is, it represents the OR of the selectors represented
// by the node selector terms.
type NodeSelector struct {
	//Required. A list of node selector terms. The terms are ORed.
	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms" protobuf:"bytes,1,rep,name=nodeSelectorTerms"`
}

// A null or empty node selector term matches no objects. The requirements of
// them are ANDed.
// The TopologySelectorTerm type implements a subset of the NodeSelectorTerm.
type NodeSelectorTerm struct {
	// A list of node selector requirements by node's labels.
	// +optional
	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,1,rep,name=matchExpressions"`
	// A list of node selector requirements by node's fields.
	// +optional
	MatchFields []NodeSelectorRequirement `json:"matchFields,omitempty" protobuf:"bytes,2,rep,name=matchFields"`
}
```

Reply:
I haven't added a test case with node affinities yet, because I don't know how I'm supposed to add them to the node itself. I'll read into it more though

Reply:
done -- added tests & made it so that selectors & affinities are `ANDed`